Mathlib/Tactic/Propose.lean

```diff
@@ -0,0 +1,129 @@
+  let cfg := if !required.isEmpty then
+    cfg.testSolutions (fun _ => do
+    let r ← instantiateMVars (.mvar orig)
+    pure <| required.all fun e => e.occurs r)
+  else
+    cfg
+  let cfg := cfg.synthInstance
+  _ ← SolveByElim.solveByElim cfg (use.toList.map pure) (pure (← getLocalHyps).toList) goals.toList
+
+/--
+Attempts to find lemmas which use all of the `required` expressions as arguments, and
+can by unified with the given `type` (which may contain metavariables, which we avoid assigning).
+We look up candidate lemmas from a discrimination tree using the first such expression.
+
+Returns an array of pairs, containing the names of found lemmas and the resulting application.
+-/
+def propose (lemmas : DiscrTree Name s) (type : Expr) (required : Array Expr)
+    (solveByElimDepth := 15) : MetaM (Array (Name × Expr)) := do
+  guard !required.isEmpty
+  let ty ← whnfR (← instantiateMVars (← inferType required[0]!))
+  let candidates ← lemmas.getMatch ty
+  candidates.filterMapM fun lem : Name =>
+    try
+      trace[Tactic.propose] "considering {lem}"
+      let Expr.mvar g ← mkFreshExprMVar type | failure
+      let e ← mkConstWithFreshMVarLevels lem
+      let (args, _, _) ← forallMetaTelescope (← inferType e)
+      let .true ← preservingMCtx <| withAssignableSyntheticOpaque <|
+        isDefEq type (← inferType (mkAppN e args)) | failure
+      g.assign (mkAppN e args)
+      let use := required.filterMap fun e => match e with | .fvar _ => none | _ => some e
+      solveByElim g (args.map fun a => a.mvarId!) use required solveByElimDepth
+      trace[Tactic.propose] "successfully filled in arguments for {lem}"
+      pure <| some (lem, ← instantiateMVars (.mvar g))
+    catch _ => pure none
+
+open Lean.Parser.Tactic
+
+/--
+* `propose using a, b, c` tries to find a lemma
+which makes use of each of the local hypotheses `a, b, c`,
+and reports any results via trace messages.
+* `propose : h using a, b, c` only returns lemmas whose type matches `h` (which may contain `_`).
+* `propose! using a, b, c` will also call `have` to add results to the local goal state.
+
+`propose` should not be left in proofs; it is a search tool, like `library_search`.
+
+Suggestions are printed as `have := f a b c`.
+-/
+syntax (name := propose') "propose" "!"? (" : " term)? (" using " (colGt term),+) : tactic
+
+open Elab.Tactic Elab Tactic in
+elab_rules : tactic |
+    `(tactic| propose%$tk $[!%$lucky]? $[ : $type:term]? using $[$terms:term],*) => do
+  let goal ← getMainGoal
+  goal.withContext do
+    let required ← terms.mapM (elabTerm · none)
+    let type ← match type with
+    | some stx => elabTermWithHoles stx none (← getMainTag) true <&> (·.1)
+    | none => mkFreshTypeMVar
+    let proposals ← propose (← proposeLemmas.get) type required
+    if proposals.isEmpty then
+      throwError "propose could not find any lemmas using the given hypotheses"
+    for p in proposals.toList.take 10 do
```

Comment:
`propose` doesn't do any reordering of the `candidates`. Would it makes sense to have `propose` stop searching once it hits `10` (configurable) results?